File: backend/dotnet/Controllers/TaskController.cs
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using TaskManager.Data;
using TaskManager.Models;

namespace TaskManager.Controllers;

[ApiController]
[Route("api")]
public class TaskController : ControllerBase
{
    private readonly ApplicationDbContext _context;
    private readonly IConfiguration _configuration;

    public TaskController(ApplicationDbContext context, IConfiguration configuration)
    {
        _context = context;
        _configuration = configuration;
    }

    private string GetBackendName()
    {
        return _configuration["BACKEND_NAME"] ?? ".NET";
    }

    [HttpGet("/")]
    public IActionResult Root()
    {
        return Ok(new { 
            message = ".NET Task Manager", 
            backend = GetBackendName(), 
            status = "running" 
        });
    }

    [HttpGet("health")]
    public IActionResult HealthCheck()
    {
        return Ok(new { 
            status = "healthy", 
            backend = GetBackendName(), 
            timestamp = DateTime.Now.ToString("yyyy-MM-ddTHH:mm:ssZ") 
        });
    }

    [HttpGet("tasks")]
    public async Task<IActionResult> GetAllTasks()
    {
        var tasks = await _context.Tasks
            .OrderByDescending(t => t.CreatedAt)
            .ToListAsync();
        
        return Ok(new { 
            success = true, 
            backend = GetBackendName(), 
            data = tasks 
        });
    }

    [HttpGet("tasks/{id}")]
    public async Task<IActionResult> GetTaskById(int id)
    {
        var task = await _context.Tasks.FindAsync(id);
        
        if (task == null)
            return NotFound(new { success = false, error = "Task not found" });
        
        return Ok(new { 
            success = true, 
            backend = GetBackendName(), 
            data = task 
        });
    }

    [HttpPost("tasks")]
    public async Task<IActionResult> CreateTask([FromBody] TaskItem taskItem)
    {
        if (string.IsNullOrEmpty(taskItem.Title))
            return BadRequest(new { success = false, error = "Title is required" });
        
        taskItem.CreatedAt = DateTime.Now;
        taskItem.Status = "pending";
        
        _context.Tasks.Add(taskItem);
        await _context.SaveChangesAsync();
        
        return CreatedAtAction(nameof(GetTaskById), new { id = taskItem.Id }, new { 
            success = true, 
            backend = GetBackendName(), 
            data = taskItem 
        });
    }

    [HttpPut("tasks/{id}")]
    public async Task<IActionResult> UpdateTask(int id, [FromBody] TaskItem taskUpdate)
    {
        var task = await _context.Tasks.FindAsync(id);
        
        if (task == null)
            return NotFound(new { success = false, error = "Task not found" });
        
        if (!string.IsNullOrEmpty(taskUpdate.Title))
            task.Title = taskUpdate.Title;
        
        if (!string.IsNullOrEmpty(taskUpdate.Description))
            task.Description = taskUpdate.Description;
        
        if (!string.IsNullOrEmpty(taskUpdate.Status))
            task.Status = taskUpdate.Status;
        
        task.UpdatedAt = DateTime.Now;
        
        await _context.SaveChangesAsync();
        
        return Ok(new { 
            success = true, 
            backend = GetBackendName(), 
            data = task 
        });
    }

    [HttpDelete("tasks/{id}")]
    public async Task<IActionResult> DeleteTask(int id)
    {
        var task = await _context.Tasks.FindAsync(id);
        
        if (task == null)
            return NotFound(new { success = false, error = "Task not found" });
        
        _context.Tasks.Remove(task);
        await _context.SaveChangesAsync();
        
        return Ok(new { 
            success = true, 
            backend = GetBackendName(), 
            message = "Task deleted successfully" 
        });
    }
}

File: backend/nodejs/src/controllers/taskController.js
const getConnection = require('../config/database');

exports.getAllTasks = async (req, res) => {
  try {
    const promisePool = await getConnection();
    const [tasks] = await promisePool.query('SELECT * FROM tasks ORDER BY created_at DESC');
    res.json({
      success: true,
      backend: process.env.BACKEND_NAME || 'Node.js',
      data: tasks
    });
  } catch (error) {
    res.status(500).json({ success: false, error: error.message });
  }
};

exports.getTaskById = async (req, res) => {
  try {
    const promisePool = await getConnection();
    const [rows] = await promisePool.query('SELECT * FROM tasks WHERE id = ?', [req.params.id]);
    const task = rows[0];
    if (!task) {
      return res.status(404).json({ success: false, error: 'Task not found' });
    }
    res.json({
      success: true,
      backend: process.env.BACKEND_NAME || 'Node.js',
      data: task
    });
  } catch (error) {
    res.status(500).json({ success: false, error: error.message });
  }
};

exports.createTask = async (req, res) => {
  try {
    const { title, description, status } = req.body;
    if (!title) {
      return res.status(400).json({ success: false, error: 'Title is required' });
    }
    
    const promisePool = await getConnection();
    const taskStatus = status || 'pending';
    const [result] = await promisePool.query(
      'INSERT INTO tasks (title, description, status) VALUES (?, ?, ?)',
      [title, description, taskStatus]
    );
    
    const [rows] = await promisePool.query('SELECT * FROM tasks WHERE id = ?', [result.insertId]);
    res.status(201).json({
      success: true,
      backend: process.env.BACKEND_NAME || 'Node.js',
      data: rows[0]
    });
  } catch (error) {
    res.status(500).json({ success: false, error: error.message });
  }
};

exports.updateTask = async (req, res) => {
  try {
    const { title, description, status } = req.body;
    const promisePool = await getConnection();
    
    const [rows] = await promisePool.query('SELECT * FROM tasks WHERE id = ?', [req.params.id]);
    if (rows.length === 0) {
      return res.status(404).json({ success: false, error: 'Task not found' });
    }
    
    await promisePool.query(
      'UPDATE tasks SET title = ?, description = ?, status = ?, updated_at = CURRENT_TIMESTAMP WHERE id = ?',
      [title || rows[0].title, description || rows[0].description, status || rows[0].status, req.params.id]
    );
    
    const [updatedRows] = await promisePool.query('SELECT * FROM tasks WHERE id = ?', [req.params.id]);
    res.json({
      success: true,
      backend: process.env.BACKEND_NAME || 'Node.js',
      data: updatedRows[0]
    });
  } catch (error) {
    res.status(500).json({ success: false, error: error.message });
  }
};

exports.deleteTask = async (req, res) => {
  try {
    const promisePool = await getConnection();
    const [rows] = await promisePool.query('SELECT * FROM tasks WHERE id = ?', [req.params.id]);
    if (rows.length === 0) {
      return res.status(404).json({ success: false, error: 'Task not found' });
    }
    
    await promisePool.query('DELETE FROM tasks WHERE id = ?', [req.params.id]);
    res.json({
      success: true,
      backend: process.env.BACKEND_NAME || 'Node.js',
      message: 'Task deleted successfully'
    });
  } catch (error) {
    res.status(500).json({ success: false, error: error.message });
  }
};

exports.healthCheck = (req, res) => {
  res.json({
    status: 'healthy',
    backend: process.env.BACKEND_NAME || 'Node.js',
    timestamp: new Date().toISOString()
  });
};